ocaml/xapi/certificates_sync.ml

```diff
@@ -35,10 +35,25 @@ let is_unchanged ~__context cert_ref cert =
   let ref_hash =
     Db.Certificate.get_fingerprint_sha256 ~__context ~self:cert_ref
   in
-  let cert_hash =
-    X509.Certificate.fingerprint `SHA256 cert |> Certificates.pp_hash
-  in
-  cert_hash = ref_hash
+  let cert_hash = Certificates.pp_fingerprint_of_cert ~hash_type:`SHA256 cert in
+  if ref_hash = "" then
+    (* We must be upgrading from a version predating fingerprint_sha256, so check fingerprint instead *)
+    if cert_hash = Db.Certificate.get_fingerprint ~__context ~self:cert_ref then (
+      let fingerprint_sha256 =
+        Certificates.pp_fingerprint_of_cert ~hash_type:`SHA256 cert
+      in
+      Db.Certificate.set_fingerprint_sha256 ~__context ~self:cert_ref
+        ~value:fingerprint_sha256 ;
+      let fingerprint_sha1 =
+        Certificates.pp_fingerprint_of_cert ~hash_type:`SHA1 cert
+      in
+      Db.Certificate.set_fingerprint_sha1 ~__context ~self:cert_ref
+        ~value:fingerprint_sha1 ;
+      true
+    ) else
+      false
+  else
+    cert_hash = ref_hash
```

Comment:
I believe this function should not be setting any fields, or doing side-effects in general.

Please keep it as-is, and add a new function that checks whether the fingerprints need to be updated, and call it just after is_unchanged is called.

Reply:
Yeah, I agree with its current name it shouldn't have side effects. However, it needs to be altered somehow. If we keep it as-is then when fingerprint_sha256 is empty it will return false and needlessly reinstall the certificate, which is what we're trying to avoid. I could change it to also return whether the fingerprints need updating, but I don't know what the function could be called as it would be returning two different things?

Reply:
Then the function that fetches from the database should be changed to return a datastructure with more information
```ocaml
type hashes = {sha256: string; sha1: string; deprecated: string}

type to_update: Certificate | Hash of hashes

val to_update : __context -> Ref.t -> X509.t -> to_update option
```
Then the caller can do a match on this variant and decide what to do:
- If the result is none, do nothing
- If it's Certificate, reinstall the certificate
- Otherwise write the collected hashes into the database